controllers/ciresource_controller_test.go

```diff
@@ -2,248 +2,168 @@ package controllers
-			cipool: cipool().name("default"),
-			cir:    cir("0").pool("default"),
-
-			waitUntil: func(cir *ofcirv1.CIResource) bool {
-				return cir.Status.State == ofcirv1.StateAvailable
-			},
-			expectedStates: []ofcirv1.CIResourceState{
-				ofcirv1.StateNone, ofcirv1.StateProvisioning, ofcirv1.StateProvisioningWait, ofcirv1.StateAvailable,
-			},
+			name: "finalizer is immediately added for a new resource",
+			testCase: newCIResourceScenario().
+				Setup(scenarioPoolWithSingleCir).
+				ReconcileUntil(func(client client.Client, cir *ofcirv1.CIResource) bool {
+					return cir.Status.State == ofcirv1.StateNone && controllerutil.ContainsFinalizer(cir, ofcirv1.OfcirFinalizer)
+				}),
 		},
 		{
-			name:   "fallback-resource-skips-provisiongwait",
-			cipool: cipool().name("fallback").priority(-1),
-			cir:    cir("0").pool("fallback"),
-
-			waitUntil: func(cir *ofcirv1.CIResource) bool {
-				return cir.Status.State == ofcirv1.StateAvailable
-			},
-			expectedStates: []ofcirv1.CIResourceState{
-				ofcirv1.StateNone, ofcirv1.StateProvisioning, ofcirv1.StateAvailable,
-			},
+			name: "new resource gets provisioned and becomes available with a valid id and address",
+			testCase: newCIResourceScenario().
+				Setup(scenarioPoolWithSingleCir).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateProvisioning
+				}).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateProvisioningWait
+				}).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateAvailable
+				}).
+				Then(func(t *testing.T, client client.Client, obj *ofcirv1.CIResource) {
+					assert.NotEmpty(t, obj.Status.ResourceId)
+					assert.NotEmpty(t, obj.Status.Address)
+				}),
 		},
 		{
-			name:   "fallback-resource-deletion",
-			cipool: cipool().name("fallback").priority(-1),
-			cir:    cir("0").pool("fallback"),
-
-			onStateChange: map[ofcirv1.CIResourceState]cirStateEvent{
-				ofcirv1.StateAvailable: func(client client.Client, cir *ofcirv1.CIResource) {
-					client.Delete(context.TODO(), cir)
-				},
-			},
-
-			waitUntil: func(cir *ofcirv1.CIResource) bool {
-				return cir == nil
-			},
+			name: "available resource can be moved under maintenance",
+			testCase: newCIResourceScenario().
+				Setup(scenarioPoolWithSingleCir).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateAvailable
+				}).
+				Then(func(t *testing.T, client client.Client, obj *ofcirv1.CIResource) {
+					obj.Spec.State = ofcirv1.StateMaintenance
+					client.Update(context.Background(), obj)
+				}).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateMaintenance
+				}),
+		},
+		{
+			name: "acquire and release an available resource",
```

Comment:
milseading name, this test only acquires, does not release

Reply:
more than a misleading name there was a missing piece :), thanks for spotting it out